### PR TITLE
egressgw: add support for excludedCIDRs

### LINKED ISSUE #23002: CFP: Egress Gateway destinationCIDRs exclusions
## Cilium Feature Proposal

**Is your feature request related to a problem?**

Egress Gateway policies are difficult to configure in the common case of wanting a default route through the gateway, but some traffic to not go through the gateway.

This is exacerbated by the the inability to support large number of CIDRs due to BPF map size limits.
https://github.com/cilium/cilium/issues/22805
https://github.com/cilium/cilium/issues/22929

For example, the following policy would be much simpler to define and would require a much smaller map if we could use a destination of `0.0.0.0/0` with a few exclusions.

```yaml
apiVersion: cilium.io/v2alpha1
kind: CiliumEgressNATPolicy
metadata:
  name: storaged-egress
spec:
  destinationCIDRs:
  - 128.0.0.0/1
  - 64.0.0.0/2
  - 32.0.0.0/4
  - 24.0.0.0/5
  - 56.0.0.0/5
  - 4.0.0.0/6
  - 12.0.0.0/6
  - 20.0.0.0/6
  - 48.0.0.0/6
  - 0.0.0.0/7
  - 8.0.0.0/7
  - 16.0.0.0/7
  - 2.0.0.0/8
  - 11.0.0.0/8
  - 19.0.0.0/8
  - 53.0.0.0/8
  - 55.0.0.0/8
  - 3.128.0.0/9
  - 10.128.0.0/9
  - 18.128.0.0/9
  - 52.0.0.0/9
  - 54.0.0.0/9
  - 3.64.0.0/10
  - 10.64.0.0/10
  - 18.64.0.0/10
  - 52.128.0.0/10
  - 54.128.0.0/10
  - 3.32.0.0/11
  - 10.32.0.0/11
  - 18.0.0.0/11
  - 52.224.0.0/11
  - 54.192.0.0/11
  - 3.16.0.0/12
  - 10.16.0.0/12
  - 18.48.0.0/12
  - 52.192.0.0/12
  - 54.240.0.0/12
  - 3.8.0.0/13
  - 10.8.0.0/13
  - 18.40.0.0/13
  - 52.208.0.0/13
  - 54.232.0.0/13
  - 3.0.0.0/14
  - 10.4.0.0/14
  - 18.36.0.0/14
  - 52.220.0.0/14
  - 54.224.0.0/14
  - 3.6.0.0/15
  - 10.0.0.0/15
  - 18.32.0.0/15
  - 52.218.0.0/15
  - 54.228.0.0/15
  - 3.4.0.0/16
  - 10.3.0.0/16
  - 18.35.0.0/16
  - 54.230.0.0/16
  - 3.5.128.0/17
  - 3.5.64.0/18
  - 18.34.64.0/18
  - 18.34.128.0/18
  - 3.5.32.0/19
  - 18.34.32.0/19
  - 18.34.192.0/19
  - 18.34.240.0/20
  - 18.34.224.0/21
  egress:
  - podSelector:
      matchExpressions:
      - key: environmentd.materialize.cloud/namespace
        operator: In
        values:
        - storage
        - compute
  egressSourceIP: 10.2.67.59
```

**Describe the feature you'd like**

The ability to specify CIDR block exclusions in the egress gateway policies.

**(Optional) Describe your proposed solution**

Add an additional field to the policy spec for `excludeCIDRs`.
Add an additional BPF map to hold the excluded CIDR blocks.
When deciding whether to route traffic through the egress gateway, also consult the map of excluded CIDR blocks.

### PR DESCRIPTION
**Pull request by jibi**
Add support to the egress gateway manager for the new excludedCIDRs
policy property.

The implementation is based on the idea of adding, for each excluded
CIDR, "zero" entries to the egress policy map (i.e. entries with the
gateway IP set to 0.0.0.0). When traffic is then matched against these
entries in the datapath, it will skip altogether all the egress gateway
logic.

Fixes: https://github.com/cilium/cilium/issues/23002

### CODE REVIEW

**Comment by jibi** [comment]
/test-only --focus="K8sDatapathEgressGatewayTest.*" --k8s_version=1.26 --kernel_version=net-next

**Comment by jibi** [comment]
/test-only --focus="K8sDatapathEgressGatewayTest.*" --k8s_version=1.26 --kernel_version=net-next

**Comment by jibi** [comment]
/test-only --focus="K8sDatapathEgressGatewayTest.*" --k8s_version=1.26 --kernel_version=net-next

**Comment by jibi** [comment]
/test-1.16-4.9

**Comment by jibi** [comment]
/test-1.24-5.4

**Comment by jibi** [comment]
/test-1.26-net-next

**Comment by benesch** [comment]
Really excited to see this, @jibi! 🎉

File: bpf/lib/egress_policies.h [hunk 1]
```diff
@@ -46,6 +46,12 @@ bool egress_gw_request_needs_redirect(struct iphdr *ip4, __u32 *tunnel_endpoint)
 	if (gateway_node_ep && (gateway_node_ep->flags & ENDPOINT_F_HOST))
 		return false;
 
+	/* If the gateway IP is 0.0.0.0 it means this is an excluded CIDR, so
+	 * skip redirection
+	 */
+	if (!egress_gw_policy->gateway_ip)
+		return false;
+
```
Comment (julianwiedmann) [inline bpf/lib/egress_policies.h]: Let's move this before the endpoint lookup :)

**Comment by jibi** [comment]
Couple of TODO items:
* add BPF unit tests. As for these I need first a generic test for redirection (we only test the SNAT path for now) based on how long it takes for this PR to get reviewed I may add them directly to this PR or open a new one
* backport https://github.com/cilium/cilium/pull/23448/commits/7b2a750edf28cb18e3dbddce7046f122ac9d310f to ensure we don't try to forward traffic to `0.0.0.0` gateways in case the agent is downgraded to an older version

**Review by nebril — APPROVED** [review]

File: pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml [hunk 2]
```diff
@@ -135,6 +135,14 @@ spec:
                 required:
                 - nodeSelector
                 type: object
+              excludedCIDRs:
+                description: excludedCIDRs is a list of destination CIDRs which will
+                  be excluded from the egress gateway redirection and SNAT logic.
+                  It should be a subset of destinationCIDRs.
```
Comment (pchaigno) [inline pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml]: What happens if it isn't?

EDIT: Reading the code, it seems that nothing will happen if it isn't. Maybe we could clarify here that it will be ignored if it isn't a subset of `destinationCIDRs`?

File: pkg/egressgateway/manager.go [hunk 3]
```diff
@@ -246,11 +247,15 @@ func (manager *Manager) addMissingIpRulesAndRoutes(isRetry bool) (shouldRetry bo
...
 
-	addIPRulesAndRoutesForConfig := func(endpointIP net.IP, dstCIDR *net.IPNet, gwc *gatewayConfig) {
+	addIPRulesAndRoutesForConfig := func(endpointIP net.IP, dstCIDR *net.IPNet, excludedCIDR bool, gwc *gatewayConfig) {
 		if !gwc.localNodeConfiguredAsGateway {
 			return
 		}
 
+		if excludedCIDR {
+			return
+		}
```
Comment (pchaigno) [inline pkg/egressgateway/manager.go]: Could you discuss the impact on those IP rules and routes in the commit description? In particular, based on these 3 lines, do they mean that some traffic that isn't matched by the policy based on exclusion CIDRs may still be matched by the IP rules?

```
$ cat ip-rule.md 
8:	from 10.20.0.67 to 35.241.162.0/24 lookup 302
```
So we may have the above even if we have an exclusion CIDR for 35.241.162.103? What's the impact of that? Why is it fine?

File: pkg/egressgateway/manager.go [hunk 4]
```diff
@@ -385,8 +399,13 @@ func (manager *Manager) removeUnusedEgressRules() {
 
 nextPolicyKey:
 	for policyKey, policyVal := range egressPolicies {
-		matchPolicy := func(endpointIP net.IP, dstCIDR *net.IPNet, gwc *gatewayConfig) bool {
-			return policyKey.Match(endpointIP, dstCIDR) && policyVal.Match(gwc.egressIP.IP, gwc.gatewayIP)
+		matchPolicy := func(endpointIP net.IP, dstCIDR *net.IPNet, excludedCIDR bool, gwc *gatewayConfig) bool {
+			gatewayIP := gwc.gatewayIP
+			if excludedCIDR {
+				gatewayIP = net.ParseIP("0.0.0.0")
```
Comment (pchaigno) [inline pkg/egressgateway/manager.go]:
```suggestion
				gatewayIP = zeroIPv4
```

File: pkg/egressgateway/policy.go [hunk 5]
```diff
@@ -159,32 +160,46 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
...
+			isExcludedCIDR := false
 			for _, dstCIDR := range config.dstCIDRs {
-				f(endpointIP, dstCIDR, &config.gatewayConfig)
+				f(endpointIP, dstCIDR, isExcludedCIDR, &config.gatewayConfig)
+			}
+
+			isExcludedCIDR = true
+			for _, excludedCIDR := range config.excludedCIDRs {
+				f(endpointIP, excludedCIDR, isExcludedCIDR, &config.gatewayConfig)
 			}
```
Comment (pchaigno) [inline pkg/egressgateway/policy.go:231]: Thanks for defining this `isExcludedCIDR`! I miss Python for that.

File: test/k8s/egress.go [hunk 6]
```diff
@@ -151,6 +150,27 @@ var _ = SkipDescribeIf(func() bool {
...
+		} else {
+			By("Check egress policy from non-gateway node")
+		}
+
+		hostIP := k8s1IP
+		if fromGateway {
+			hostIP = k8s2IP
+		}
+
+		src, _ := fetchPodsWithOffset(kubectl, randomNamespace, "client", testDSClient, hostIP, false, 1)
```
Comment (pchaigno) [inline test/k8s/egress.go]: Nit: Couldn't we merge this with `testEgressGateway` above with a flag to decide what case we're checking?

File: pkg/egressgateway/manager.go [hunk 7]
```diff
@@ -305,9 +305,15 @@ nextIpRule:
...
-				ipRule.Src.IP.Equal(endpointIP) && ipRule.Dst.String() == dstCIDR.String()
+			if !manager.installRoutes {
+				return false
+			}
+
+			if !gwc.localNodeConfiguredAsGateway {
+				return false
+			}
+
+			return ipRule.Src.IP.Equal(endpointIP) && ipRule.Dst.String() == dstCIDR.String()
```
Comment (pchaigno) [inline pkg/egressgateway/manager.go:329]: Nice cleanup!

**Review by pchaigno — APPROVED** [review]
Nice pull request (as usual :smiley:)!

My only concern is with the ip rules and routes as I'm not sure I understand the impact there.

Regarding downgrade, should we ensure we get the commit into v1.13 before we merge this? Or alternatively, create a release blocker issue to ensure this PR doesn't end up in v1.14 before v1.13 is ready (in case something comes up and we forget to backport that commit)?
On the same topic, is this something we want in v1.13? (In which case we'll have to backport the commit to v1.12.)

File: pkg/egressgateway/policy.go (hunk 5 above)
Comment (jibi) [reply pkg/egressgateway/policy.go:231]: (or Rust :disappear:)

File: pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml (hunk 2 above)
Comment (jibi) [reply pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml]: Correct, this was more like "to be effective it must be a subset" but nothing bad happens otherwise, let me add that :+1:

File: test/k8s/egress.go (hunk 6 above)
Comment (jibi) [reply test/k8s/egress.go]: Pushed https://github.com/cilium/cilium/pull/23448/commits/113ea0e834cfb59c7951cd276d3365d26eb3b677 to ease the review (and test it), will squash it with the manager changes once we are happy (tests are green https://jenkins.cilium.io/job/Cilium-PR-Tests-Kernel-Focus/522/)

**Comment by jibi** [comment]
/test-only --focus="K8sDatapathEgressGatewayTest.*" --k8s_version=1.26 --kernel_version=net-next

File: pkg/egressgateway/manager.go (hunk 3 above)
Comment (jibi) [reply pkg/egressgateway/manager.go]: > Why is it fine?

Nice catch, no, it's not fine as we need to exclude those CIDRs also from the IP rules/routes :disappointed: 

Need to think what's the best way to tackle this

**Comment by jibi** [comment]
/test-only --focus="K8sDatapathEgressGatewayTest.*" --k8s_version=1.26 --kernel_version=net-next

**Comment by jibi** [comment]
/test-only --focus="K8sDatapathEgressGatewayTest.*" --k8s_version=1.26 --kernel_version=net-next

**Comment by jibi** [comment]
/test-only --focus="K8sDatapathEgressGatewayTest.*" --k8s_version=1.26 --kernel_version=net-next

**Comment by jibi** [comment]
/test-only --focus="K8sDatapathEgressGatewayTest.*" --k8s_version=1.26 --kernel_version=net-next

**Comment by jibi** [comment]
Updates since the first iteration:
* added some more BPF and manager unit tests
* refactored the e2e tests as suggested
* fixed the IP rules/routes logic (see https://github.com/cilium/cilium/pull/23448/commits/42698e17614e0e2742626993f12c36decf001381 for more context)

File: pkg/egressgateway/policy.go [hunk 8]
```diff
@@ -177,33 +179,114 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
...
+			}
+
+			dstCIDRMinusExcludedCIDRs = newDstCIDRMinuxExcludedCIDRs
+		}
+
+		cidrs = append(cidrs, dstCIDRMinusExcludedCIDRs...)
+	}
+
+	return cidrs
+}
```
Comment (pchaigno) [inline pkg/egressgateway/policy.go:207]: Nit: This function alone would probably warrant its own commit.

I ran this on small examples and tried to turn it around. It looks good to me and I'm not seeing a more efficient implementation (unfortunately). We will have a lot of cases where we attempt to partition two CIDRs that don't overlap but luckily `PartitionCIDR` has special cases to handle that quickly.

That said, we probably don't want to call this monster too often.

File: pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml [hunk 9]
```diff
@@ -135,6 +135,15 @@ spec:
                 required:
                 - nodeSelector
                 type: object
+              excludedCIDRs:
+                description: excludedCIDRs is a list of destination CIDRs which will
```
Comment (nathanjsweet) [inline pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml]:
```suggestion
                description: excludedCIDRs is a list of destination CIDRs that will
```

File: pkg/k8s/apis/cilium.io/v2/cegp_types.go [hunk 10]
```diff
@@ -52,6 +52,14 @@ type CiliumEgressGatewayPolicySpec struct {
 	// If a destination IP matches any one CIDR, it will be selected.
 	DestinationCIDRs []IPv4CIDR `json:"destinationCIDRs"`
 
+	// excludedCIDRs is a list of destination CIDRs which will be excluded
```
Comment (nathanjsweet) [inline pkg/k8s/apis/cilium.io/v2/cegp_types.go]:
```suggestion
	// ExcludedCIDRs is a list of destination CIDRs that will be excluded
```

**Review by nathanjsweet — CHANGES_REQUESTED** [review]
For API, a couple of nits.

**Review by nathanjsweet — APPROVED** [review]

File: pkg/egressgateway/manager.go [hunk 11]
```diff
@@ -27,7 +27,8 @@ import (
 )
```
Comment (pchaigno) [inline pkg/egressgateway/manager.go:28]: > Moreover, to make IP rules and routes work with excluded CIDRs, ...

It would really help reviewing if the IP routes and rules changes would be in their own commit. Right now, this commit includes the basic support, the IP rules & routes handling, and all the helper functions. When reading the helper functions in particular, the lack of separation makes it even harder to understand why/how they are used.

File: pkg/egressgateway/policy.go [hunk 12]
```diff
@@ -177,33 +179,114 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
...
+
+			isExcludedCIDR = true
+			for _, excludedCIDR := range config.excludedCIDRs {
+				f(endpointIP, excludedCIDR, isExcludedCIDR, &config.gatewayConfig)
+			}
+		}
+	}
+}
+
+func (config *PolicyConfig) forEachEndpointAndDestinationMinusExcludedCIDRs(epDataStore map[endpointID]*endpointMetadata,
```
Comment (pchaigno) [inline pkg/egressgateway/policy.go]: What about `forEachEndpointAndDestination` becomes `forEachEndpointAndCIDR` while the new function becomes `forEachEndpointAndComputedDestination` (or just `forEachEndpointAndDestination` if that's too long). The rationale being that the first doesn't really process "destinations" anymore but two CIDR lists; the second only processes destination because we computed the diff.

In any case, you need comments on all of those helper functions to explain when it uses the input CIDR lists as is vs. the diff of the input CIDRs.

File: pkg/egressgateway/manager.go [hunk 13]
```diff
@@ -27,7 +27,8 @@ import (
...
```
Comment (pchaigno) [inline pkg/egressgateway/manager.go:27]: > * forEachEndpointAndDestinationMinusExcludedCIDRs: works like
  forEachEndpointAndDestination, but iterate over
  destinationCIDRs - excludedCIDRs
> * matchesMinusExcludedCIDRs: works like matches but matches over
  destinationCIDRs - excludedCIDRs

Please replace `destinationCIDRs - excludedCIDRs` by actual words. Maybe I'm tired, but I didn't understand what you meant before I read the code. `but matches over the diff of destinationCIDRs minus excludedCIDRs` would probably be enough.

File: Documentation/network/egress-gateway.rst [hunk 14]
```diff
@@ -205,6 +205,20 @@ One or more IPv4 destination CIDRs can be specified with ``destinationCIDRs``:
...
+
+.. code-block:: yaml
+
+    destinationCIDRs:
+    - "a.b.0.0/16"
+    excludedCIDRs:
+    - "a.b.c.0/24"
+
+In this case traffic destined to the ``a.b.0.0/16`` CIDR, except for the
+``a.b.c.0/24`` destination, will go through egress gateway and leave the cluster
```
Comment (pchaigno) [inline Documentation/network/egress-gateway.rst:223]: Maybe we should make the two CIDRs completely unrelated in the example to not give the wrong impression that the excludedCIDR needs to be a subset of the destinationCIDR?

**Review by pchaigno — APPROVED** [review]
Only minor comments from me, but the main commit (`egressgw: add support for excludedCIDRs`) could use some work to make it easier to follow. It's actually fairly simple changes if you already understand the main idea, but it's currently still hard to follow as is.

File: pkg/egressgateway/policy.go [hunk 15]
```diff
@@ -177,33 +179,114 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
...
+	cidr := config.destinationMinusExcludedCIDRs()
+
+	for _, endpoint := range epDataStore {
+		if !config.selectsEndpoint(endpoint) {
+			continue
+		}
+
+		for _, endpointIP := range endpoint.ips {
+			for _, cidr := range cidr {
+				if f(endpointIP, cidr, &config.gatewayConfig) {
```
Comment (julianwiedmann) [inline pkg/egressgateway/policy.go]: I'm amazed by the `cidr := range cidr`. Can we give these two variables different names?

File: Documentation/network/egress-gateway.rst (hunk 14 above)
Comment (qmonnet) [reply Documentation/network/egress-gateway.rst:223]: Wrong impression? My understanding was the opposite?

```
              excludedCIDRs:
                description: ExcludedCIDRs is a list of destination CIDRs that will
                  be excluded from the egress gateway redirection and SNAT logic.
                  Should be a subset of destinationCIDRs otherwise it will not have
                  any effect.
```

File: pkg/egressgateway/manager_privileged_test.go [hunk 16]
```diff
@@ -287,8 +403,15 @@ func newCiliumNode(name, nodeIP string, nodeLabels map[string]string) nodeTypes.
...
-	_, destCIDR, _ := net.ParseCIDR(destinationCIDR)
+func newEgressPolicyConfigWithNodeSelector(policyName string, labels map[string]string, destinationCIDR string, excludedCIDRs []string, selector *v1.LabelSelector, iface string) PolicyConfig {
+	_, parsedDestinationCIDR, _ := net.ParseCIDR(destinationCIDR)
+
+	parsedExcludedCIDRs := []*net.IPNet{}
+	for _, excludedCIDR := range excludedCIDRs {
+		_, parsedExcludedCIDR, _ := net.ParseCIDR(excludedCIDR)
+		parsedExcludedCIDRs = append(parsedExcludedCIDRs, parsedExcludedCIDR)
+
+	}
```
Comment (qmonnet) [inline pkg/egressgateway/manager_privileged_test.go]: Nit: blank line.

**Review by qmonnet — APPROVED** [review]
Looks good to me, thanks!

File: pkg/egressgateway/policy.go (hunk 12 above)
Comment (jibi) [reply pkg/egressgateway/policy.go]: > What about forEachEndpointAndDestination becomes forEachEndpointAndCIDR while the new function becomes forEachEndpointAndComputedDestination (or just forEachEndpointAndDestination if that's too long). The rationale being that the first doesn't really process "destinations" anymore but two CIDR lists; the second only processes destination because we computed the diff.

The intent here was to have a self explanatory name, so that I don't have to go and look it up in the comment what it's doing: these are `PolicyConfig`'s methods, so the `Endpoint` and `Destination` part of `forEachEndpointAndDestination` simply refers to the policy's fields:
https://github.com/cilium/cilium/blob/5510a770cd5c013492df582fcd4caef0eae42e79/pkg/egressgateway/policy.go#L60
https://github.com/cilium/cilium/blob/5510a770cd5c013492df582fcd4caef0eae42e79/pkg/egressgateway/policy.go#L61

(so to me it reads like "iterate over all combinations of EP and destination CIDRs").

Same idea for the other `forEachEndpointAndDestinationMinusExcludedCIDRs` version, which although very mouthful, IMO makes it more clear what we are iterating over: EPs and `dstCIDRs` slice minus `excludedCDIRs` slice.

Perhaps for consistency I can rename the first one to `forEachEndpointAndDestinationCIDRs`

File: pkg/egressgateway/policy.go (hunk 15 above)
Comment (jibi) [reply pkg/egressgateway/policy.go]: I blame `:s/<something>/<something>/g` (the returned slice from `config.destinationMinusExcludedCIDRs()` should really be `cidrs`)

File: Documentation/network/egress-gateway.rst (hunk 14 above)
Comment (jibi) [reply Documentation/network/egress-gateway.rst:223]: That's on purpose, excludedCIDRs needs indeed to be a subset of destinationCIDRs, otherwise it will have no effect

File: pkg/egressgateway/policy.go (hunk 12 above)
Comment (pchaigno) [reply pkg/egressgateway/policy.go]: > The intent here was to have a self explanatory name, so that I don't have to go and look it up in the comment what it's doing: these are `PolicyConfig`'s methods, so the `Endpoint` and `Destination` part of `forEachEndpointAndDestination` simply refers to the policy's fields:

I think this is too complex to hope that the name can be self-explanatory. I don't think you can avoid adding a comment for those helper functions.

One issue with `forEachEndpointAndDestinationMinusExcludedCIDRs` is that it's unclear, without reading the function itself, if it iterates over all endpoints and destinations, but doesn't consider excluded CIDRs ("minus/except excluded CIDRs") or if it iterates over all endpoints and diffed destinations ("minus/diffed by excluded CIDRs").

Comment (jibi) [reply pkg/egressgateway/policy.go]: Fair enough, I'll add some comments then :+1:

File: Documentation/network/egress-gateway.rst (hunk 14 above)
Comment (pchaigno) [reply Documentation/network/egress-gateway.rst:223]: I think I was confused on how CIDRs work :sweat_smile: If two CIDRs overlap, then they are either equal or one is the subset of the other.

File: pkg/egressgateway/policy.go (hunk 12 above)
Comment (jibi) [reply pkg/egressgateway/policy.go]: Scratch that, after taking another fresh look I'm also convinced that `forEachEndpointAndCIDR` and `forEachEndpointAndDestination` are better names, so I went for them :+1:

Comment (pchaigno) [reply pkg/egressgateway/policy.go]: Not to be a pain, but I think a quick comment would still help :slightly_smiling_face:

Comment (jibi) [reply pkg/egressgateway/policy.go]: Yes, I added them already in the latest push

**Comment by jibi** [comment]
/test-1.16-4.19

**Review by julianwiedmann — APPROVED** [review]
:rocket: :rocket: :rocket:

**Comment by jibi** [comment]
/test-1.16-4.19
